cpp/src/parquet/properties.cc

```diff
@@ -0,0 +1,42 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <utility>
+
+#include "parquet/properties.h"
+
+#include "arrow/io/buffered.h"
+#include "arrow/io/memory.h"
+
+namespace parquet {
+
+std::shared_ptr<ArrowInputStream> ReaderProperties::GetStream(
+    std::shared_ptr<ArrowInputFile> source, int64_t start, int64_t num_bytes) {
+  if (buffered_stream_enabled_) {
+    std::shared_ptr<::arrow::io::BufferedInputStream> stream;
+    PARQUET_THROW_NOT_OK(source->Seek(start));
+    PARQUET_THROW_NOT_OK(::arrow::io::BufferedInputStream::Create(
+        buffer_size_, pool_, source, &stream, num_bytes));
```

Comment:
The `ArrowInputFile` is a `RandomAccessFile`. However, `::arrow::io::BufferedInputStream` takes an `InputStream`. Casting `RandomAccessFile` to `InputStream` is incorrect since the `::arrow::io::BufferedInputStream::Peek` causes the `RandomAccessFile` offset to change.

Reply:
Generally speaking, when using a buffering layer, the buffering layer owns the underlying raw file. Using both at once is incorrect. This is not Arrow-specific, but happens in any language (Python, C, etc.).

Reply:
This requirement was not explicitly spelled out before, it only behaved as you wanted "by accident". Let's open a new JIRA and see what can be done